### PR TITLE
[AOTInductor] Remove call to aot_autograd when receiving ExportedProgram

### PR DESCRIPTION
**Pull request by angelayi**
https://github.com/pytorch/pytorch/issues/105555

Existing flow first exports and then calls torch._inductor.aot_compile. However, export calls aot_autograd with the core aten decomposition table, and then torch._inductor.aot_compile calls aot_autograd again with the inductor decomposition table. The 2nd calling of aot_autograd is supposedly causing some problems, and seems excessive, so instead we will create a new function, torch._export.aot_compiler which will export using the inductor decomposition table, pass it to inductor's compile_fx_aot, and because it has already been exported, avoid recalling aot_autograd.

```
def aot_compile(
    f: Callable,
    args: Tuple[Any],
    kwargs: Optional[Dict[str, Any]] = None,
    constraints: Optional[List[Constraint]] = None,
) -> Tuple[str, ExportedProgram]:
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

### CODE REVIEW

File: torch/_export/_experimental.py
Comment (angelayi) [inline torch/_export/_experimental.py:10]: We can't have torch._inductor.aot_compile take in an ExportedProgram because of circular import errors...
But another solution is that because we're in python we could also just ignore the annotations and pass torch._inductor.aot_compile an ExportedProgram and import + check inside that function.

File: torch/_inductor/config.py
Comment (desertfire) [inline torch/_inductor/config.py:68]: I would prefer to make AOTInductor takes exported program as a mandatory requirement, with some checking in comple_fx_aot. wrapper_call in AotCodeCache is too hacky and not reliable anyway, and I think giving user too much flexibility as in the old way isn't necessarily a good choice.

File: torch/_export/__init__.py
Comment (desertfire) [inline torch/_export/__init__.py:156]: Do we want to keep this only for AOTInductor? Also when decomposition_table is not None, you want to merge the two dicts.

File: torch/_inductor/compile_fx.py
Comment (desertfire) [inline torch/_inductor/compile_fx.py:232]: This simple materialize was causing problem for some models previously. Do you think the inputs will all be real tensors in the export+aot_compile path? If yes, we should be fine here.

Comment (angelayi) [reply torch/_inductor/compile_fx.py:232]: I think the original code also expects the inputs to always be real tensors? The original code differs by just getting the parameters from the tracing context, but from export the parameters are already passed in as inputs.

**Comment by desertfire** [comment]
Please also remember to update https://github.com/pytorch/pytorch/blob/78fffe890627bc020c9bd82a27a19d042f31695e/benchmarks/dynamo/common.py#L1090

**Comment by angelayi** [comment]
After elias lands his diff, I can rebase, import to fbsource, and fix internal use cases.

File: torch/_export/__init__.py
Comment (desertfire) [reply torch/_export/__init__.py:156]: `dict.update` can take precedence, but I guess arguably either way is fine. Just need to make the behavior clear in the user API comment/doc.

File: test/inductor/test_aot_inductor.py
Comment (desertfire) [inline test/inductor/test_aot_inductor.py:35]: `select_decomp_table()` may be better as it respects `config.fallback_random`.

File: benchmarks/dynamo/common.py
Comment (desertfire) [inline benchmarks/dynamo/common.py:1112]: pass in `decomposition_table`?

File: torch/_export/__init__.py
Comment (angelayi) [reply torch/_export/__init__.py:156]: Hmm, but what if a user removed something from the core aten decomp table? Then by adding it, it'll get decomposed again.

File: torch/_inductor/graph.py
Comment (desertfire) [inline torch/_inductor/graph.py:941]: This will always be true because of `if self.aot_mode and self.cpp_wrapper`? I think can just get rid of `AotCodeCache.compile`.

**Review by desertfire — APPROVED** [review]
CI failure is related to this file, https://github.com/pytorch/pytorch/blob/main/test/cpp/aot_inductor/test.py, which needs to be updated.

Approve to unblock, need to fix small issues before landing.

File: benchmarks/dynamo/common.py
Comment (desertfire) [inline benchmarks/dynamo/common.py:1114]: Also we probably want to provide a specific API to target Export+AOTInductor. It's awkward for most end users to figure out which decomp table to pass in here.

Comment (desertfire) [reply benchmarks/dynamo/common.py:1114]: Ok for this PR. We can address it later.

File: torch/_export/__init__.py
Comment (zhxchen17) [inline torch/_export/__init__.py:130]: remove these?

File: torch/_export/__init__.py
Comment (zhxchen17) [reply torch/_export/__init__.py:157]: I guess from API design point of view you can do two things as Bin suggested:
1. overwrite
2. merge
overwrite is fine but the downside is user is implicitly depending on the fact that export() uses core aten IR by default, which is a little bit leaky abstraction to me, but I guess in practice it's fine?

If we wanna go with merge route, then we need an "addition set" and a "minus set" which are two arguments.

**Comment by zhxchen17** [comment]
export() part has some major nits need to be addressed, otherwise looks fine overall

File: torch/_export/__init__.py
Comment (angelayi) [reply torch/_export/__init__.py:130]: Oh, somehow they were there...

File: benchmarks/dynamo/common.py
Comment (angelayi) [reply benchmarks/dynamo/common.py:1114]: Makes sense -- do you have an opinion on where this API should live? In export or in inductor?

File: torch/_export/__init__.py
Comment (angelayi) [reply torch/_export/__init__.py:157]: My preference is to go with overwriting, but another idea is, what if we use a context manager to specify the decomp table rather having it as an argument to export? 

My feeling is that the manipulation of the decomp table will be used by super users, so if they do want to specify their own decompositions then they can specify it through a context manager like:
```
with set_export_decomp_table(inductor_decomp):
    ep = export(...)
```

Comment (zhxchen17) [reply torch/_export/__init__.py:154]: I think it's still leaky whether you make it context manager or not. User just express their intention differently here:
1. For overwrite, user basically says: no matter what export() does, use MY decomp table.
2. For merge, use says: on top of export() decomp behavior, I want these to be decomped and those to be not decomped. 

I think 1 and 2 are not the same although in practice they could achieve the same thing. I think context manager v.s. argument is just implicit vs explicit over an API which is orthogonal to the semantics difference I mentioned above.

File: benchmarks/dynamo/common.py
Comment (angelayi) [reply benchmarks/dynamo/common.py]: Added the API to torch._export

File: torch/_export/__init__.py
Comment (angelayi) [reply torch/_export/__init__.py]: Chatted offline: since we don't have a strong use case for passing in the decomp table yet, we will just patch the decomp table directly in our helper function

**Review by zhxchen17 — APPROVED** [review]

File: torch/_inductor/codecache.py
Comment (eellison) [inline torch/_inductor/codecache.py:892]: Why do we need these changes ?

File: torch/_inductor/compile_fx.py
Comment (eellison) [inline torch/_inductor/compile_fx.py:223]: could you add a comment what's happening here ?

also, this is my bad, but while you're here would you mind refomating the 
```
                    real_inputs = [
                        materialize(x)
                        for x in [
                            *[
                                param
                                for param in torch._guards.TracingContext.get().params_flat
                                if param is not None
                            ],
                            *V.real_inputs,
                        ]
                    ]
```

lines

File: torch/_inductor/compile_fx.py
Comment (eellison) [inline torch/_inductor/compile_fx.py:540]: just for my own understanding, why do we need this ?

File: torch/_inductor/compile_fx.py
Comment (eellison) [inline torch/_inductor/compile_fx.py:1092]: This is skipping freezing because we changed it to invoking `fw_compiler` instead of `inference_compiler`

**Review by eellison — CHANGES_REQUESTED** [review]
we're not invoking freezing currently

File: torch/_inductor/codecache.py
Comment (angelayi) [reply torch/_inductor/codecache.py:892]: Instead of returning a function, @desertfire said we should just directly return the so_path, so this just gets rid of the wrapper.

File: torch/_inductor/compile_fx.py
Comment (angelayi) [reply torch/_inductor/compile_fx.py:540]: My understanding is, if we're just aot compiling, we do not need to construct the followup `CompiledFXGraph` and we can just return the "compiled_fn" which in this case is just the so_path from AOTCodeCache

File: torch/_inductor/compile_fx.py
Comment (angelayi) [reply torch/_inductor/compile_fx.py:1092]: I have a followup PR to change this to `inference_compiler`: https://github.com/pytorch/pytorch/pull/106239, but right now freezing doesn't work with the c++ wrapper so I wasn't sure if I should make that change here.

Comment (eellison) [reply torch/_inductor/compile_fx.py:1092]: sounds good

**Review by eellison — APPROVED** [review]

**Comment by kimishpatel** [comment]
Angela, can you say, and perhaps add to summary, what this PR is doing?

**Pull request closed without merging**
